polaris/experimental/_benchmark_v2.py

```diff
@@ -0,0 +1,247 @@
+        return len(self.indices)
+
+    @computed_field
+    @cached_property
+    def md5_checksum(self) -> str:
+        return md5(self.serialize()).hexdigest()
+
+    def intersect(self, other: "IndexSet") -> bool:
+        return self.indices.intersect(other.indices)
+
+    def serialize(self) -> bytes:
+        return self.indices.serialize()
+
+    @staticmethod
+    def deserialize(index_set: bytes) -> "IndexSet":
+        return IndexSet(indices=BitMap.deserialize(index_set))
+
+
+class SplitV2(BaseModel):
+    training: IndexSet
+    test: IndexSet
+
+    @field_validator("training", "test", mode="before")
+    @classmethod
+    def _parse_index_sets(cls, v: bytes | IndexSet) -> bytes | IndexSet:
+        """
+        Accepted a binary serialized IndexSet
+        """
+        if isinstance(v, bytes):
+            return IndexSet.deserialize(v)
+        return v
+
+    @field_validator("training")
+    @classmethod
+    def _validate_training_set(cls, v: IndexSet) -> IndexSet:
+        """
+        Training index set can be empty (zero-shot)
+        """
+        if v.datapoints == 0:
+            logger.info(
+                "This benchmark only specifies a test set. It will return an empty train set in `get_train_test_split()`"
+            )
+        return v
+
+    @field_validator("test")
+    @classmethod
+    def _validate_test_set(cls, v: IndexSet) -> IndexSet:
+        """
+        Test index set cannot be empty
+        """
+        if v.datapoints == 0:
+            raise InvalidBenchmarkError("The predefined split contains empty test partitions")
+        return v
+
+    @model_validator(mode="after")
+    def validate_set_overlap(self) -> Self:
+        """
+        The training and test index sets do not overlap
+        """
+        if self.training.intersect(self.test):
+            raise InvalidBenchmarkError("The predefined split specifies overlapping train and test sets")
+        return self
+
+    @property
+    def n_train_datapoints(self) -> int:
+        """
+        The size of the train set.
+        """
+        return self.training.datapoints
+
+    @property
+    def n_test_sets(self) -> int:
+        """
+        The number of test sets
+        """
+        # TODO: Until we support multi-test benchmarks
+        return 1
+
+    @property
+    def n_test_datapoints(self) -> dict[str, int]:
+        """
+        The size of (each of) the test set(s).
+        """
+        return {"test": self.test.datapoints}
+
+    @property
+    def max_index(self) -> int:
+        return max(self.training.indices.max(), self.test.indices.max())
+
+    def test_items(self) -> Generator[tuple[str, IndexSet], None, None]:
+        # TODO: Until we support multi-test benchmarks
+        yield "test", self.test
+
+
+class BenchmarkV2Specification(BenchmarkSpecification):
+    _version: ClassVar[Literal[2]] = 2
+
+    dataset: DatasetV2 = Field(exclude=True)
+    split: SplitV2
+    n_classes: dict[ColumnName, int]
```

Comment:
**Question:** Since `n_classes` is in both the V1 and V2, could there be a way to define it in the base class? Or is that hard to do given that the implementations differ (i.e. `computed_property` vs. a regular field)

Reply:
That's technically possible, but it would add a lot of boilerplate.